Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,129 @@
+
+
+## 3. Audience
+
+Users with use cases where scraped load is very large or hard to estimate and/or scalability is hard to predict, so they need a simple way to manage load distribution and scalability. They may apply DaemonSet deployment on the whole cluster or only some nodes according to their needs.
+
+An example of audience is expressed [here](https://github.com/prometheus-operator/prometheus-operator/issues/5495#issuecomment-1519812510).
+
+
+## 4. Goals
+
+Provide an MVP version of the DaemonSet deployment of Prometheus Agent to the Audience.
+In specific, the MVP will need to:
+* Allow users to deploy one Prometheus Agent pod per node.
+* Allow users to restrict which set of nodes they want to deploy Prometheus Agent, if desired.
+* Allow users to set the priority of Prometheus Agent pod compared to other pods on the same node, if desired.
+* Allow each Prometheus Agent pod to only scrape from the pods from PodMonitor that run on the same node.
+
+
+## 5. Non-Goals
+
+This proposal only aims at Prometheus Agent, not Prometheus.
+
+Other non-goals are the features that are not easy to implement and require more investigation. We will need to investigate whether there are actual user needs for them, if yes, then how to best implement them. We can also learn from similar projects such as OpenTelemetry Operator (they have DaemonSet mode) and Grafana Agent on how they approach these problems and what we can apply for our cases. We’ll handle these after the MVP. Those (currently) non-goals features are:
+* ServiceMonitor support: There's a performance issue regarding this feature. Since each Prometheus Agent running on a node requires one watch, making all Prometheus Agent pods watch all endpoints will put a huge stress on Kubernetes API server. This is the main reason why GMP hasn’t supported this, even though there are user needs stated in some issues ([#362](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/362), [#192](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192)). However, as discussed with Danny from GMP [here](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192#issuecomment-2028850846), ServiceMonitor support based on EndpointSlice seems like a viable approach. We’ll investigate this further after the MVP.
+* Storage: We will need to spend time studying more about the WAL, different storage solutions provided by Kubernetes, and how to gracefully handle storage in different cases of crashes. For example, there’s an [issue in Prometheus](https://github.com/prometheus/prometheus/issues/8809) showing that samples may be lost if remote write didn’t flush cleanly. We’ll investigate these further after the MVP.
+
+In the MVP version, we will not allow users to directly switch from a live StatefulSet to DaemonSet deployment. Reasons are explained in the CRD subsection in the How section.
+
+
+## 6. How
+
+The MVP version of DaemonSet deployment will be put behind a feature gate.
```

Comment:
```suggestion
The MVP version of DaemonSet deployment will be put behind a feature gate until we have enough confidence that the functionality is ready for testing by the upstream community at large. It will allow us to develop the feature in lock steps while maintaining the stability of the next releases.
```